### PR TITLE
added indexregistry

### PR DESCRIPTION
**Pull request by Sidu28**
Adds a separate registry for managing operators position in a list of operatorsIds both globally and on a quorum-by-quorum basis

### CODE REVIEW

File: src/contracts/interfaces/IIndexRegistry.sol [hunk 1]
```diff
@@ -31,10 +31,11 @@ interface IIndexRegistry is IRegistry {
      * @notice Deregisters the operator with the specified `operatorId` for the quorums specified by `quorumBitmap`.
      * @param operatorId is the id of the operator that is being deregistered
      * @param quorumNumbers is the quorum numbers the operator is deregistered for
-     * @param indexes is an array of indexes for each quorum as witnesses for the last operators to swap for each quorum
+     * @param quorumToOperatorListIndexes is an array of indexes for each quorum as witnesses for the last operators to swap for each quorum
+     * @param globalOperatorListIndex is the index of the operator that is to be removed from the list
      * @dev access restricted to the RegistryCoordinator
      */
-    function deregisterOperator(bytes32 operatorId, uint8[] memory quorumNumbers, uint32[] memory indexes) external;
+    function deregisterOperator(bytes32 operatorId, uint8[] memory quorumNumbers, uint32[] memory quorumToOperatorListIndexes, uint32 globalOperatorListIndex) external;
```
Comment (gpsanant) [inline src/contracts/interfaces/IIndexRegistry.sol:38]: can you use the function signatures in this interface: https://github.com/Layr-Labs/eigenlayer-contracts/blob/impl-registry-coordinator/src/contracts/interfaces/IIndexRegistry.sol

main change is quorumNumbers is bytes memory

File: src/contracts/middleware/IndexRegistry.sol [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        registryCoordinator = _registryCoordinator;
+    }
+
+    /**
+     * @notice Registers the operator with the specified `operatorId` for the quorums specified by `quorumBitmap`.
+     * @param operatorId is the id of the operator that is being registered
+     * @param quorumNumbers is the quorum numbers the operator is registered for
+     * @dev access restricted to the RegistryCoordinator
+     */
+    function registerOperator(bytes32 operatorId, uint8[] memory quorumNumbers) external onlyRegistryCoordinator {
```
Comment (gpsanant) [inline src/contracts/middleware/IndexRegistry.sol:41]: Can you add the preconditions that you think are assuming here? example: https://github.com/Layr-Labs/eigenlayer-contracts/blob/d69f30ec857d525b8452d85c6549eee6057bad22/src/contracts/interfaces/IStakeRegistry.sol#L30

File: src/contracts/middleware/IndexRegistry.sol [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        //add operator to operatorList
+        globalOperatorList.push(operatorId);
+
+        for (uint i = 0; i < quorumNumbers.length; i++) {
+            quorumToOperatorList[quorumNumbers[i]].push(operatorId);
```
Comment (gpsanant) [inline src/contracts/middleware/IndexRegistry.sol:46]: why do we need a list per quorum? Could we just keep track the indexes in the other storage scheme we chatted about?

File: src/contracts/middleware/IndexRegistry.sol [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    /// @notice Returns the current number of operators of this service.
+    function totalOperators() external view returns (uint32){
+        return uint32(globalOperatorList.length);
+    }
+
+
+    function _updateTotalOperatorHistory(uint8 quorumNumber) internal {
+
+        //if there is a prior entry, update its "toBlockNumber"
+        if (totalOperatorsHistory[quorumNumber].length > 0) {
```
Comment (gpsanant) [inline src/contracts/middleware/IndexRegistry.sol:130]: we prematurely load the length and store in memory in other places, we should do it here too

File: src/contracts/middleware/IndexRegistry.sol [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    /// @param indexToRemove index of the operator to remove
+    function _removeOperatorFromQuorumToOperatorList(uint8 quorumNumber, uint32 indexToRemove) internal {   
+        
+        uint32 quorumToOperatorListLastIndex  = uint32(quorumToOperatorList[quorumNumber].length - 1);
+
+        // if the operator is not the last in the list, we must swap the last operator into their positon
+        if(indexToRemove != quorumToOperatorListLastIndex){
+            bytes32 operatorIdToSwap = quorumToOperatorList[quorumNumber][quorumToOperatorListLastIndex];
+            //update the swapped operator's operatorIdToIndexHistory list with a new entry, as their index has now changed
+            _updateOperatorIdToIndexHistory(operatorIdToSwap, quorumNumber, indexToRemove);
```
Comment (gpsanant) [inline src/contracts/middleware/IndexRegistry.sol:168]: we also need to update the removed operator's index history by setting the `toBlockNumber` to the current block.

File: src/contracts/middleware/IndexRegistry.sol [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+     * @param quorumToOperatorListIndexes is an array of indexes for each quorum as witnesses for the last operators to swap for each quorum
+     * @param globalOperatorListIndex is the index of the operator that is to be removed from the list
+     * @dev access restricted to the RegistryCoordinator
+     */
+    function deregisterOperator(bytes32 operatorId, uint8[] memory quorumNumbers, uint32[] memory quorumToOperatorListIndexes, uint32 globalOperatorListIndex) external onlyRegistryCoordinator {
+        require(quorumNumbers.length == quorumToOperatorListIndexes.length, "IndexRegistry.deregisterOperator: quorumNumbers and indexes must be the same length");
+        _removeOperatorFromGlobalOperatorList(globalOperatorListIndex);  
+
+        for (uint i = 0; i < quorumNumbers.length; i++) {
+            _removeOperatorFromQuorumToOperatorList(quorumNumbers[i], quorumToOperatorListIndexes[i]);
```
Comment (gpsanant) [inline src/contracts/middleware/IndexRegistry.sol:65]: where is the validation being done on these indices?

File: src/contracts/middleware/IndexRegistry.sol (hunk 5 above)
Comment (Sidu28) [reply src/contracts/middleware/IndexRegistry.sol:168]: https://github.com/Layr-Labs/eigenlayer-contracts/blob/a0381d062e2d64bd8d45088b6f2281b3b99ed01d/src/contracts/middleware/IndexRegistry.sol#L68

File: src/contracts/middleware/IndexRegistry.sol (hunk 4 above)
Comment (Sidu28) [reply src/contracts/middleware/IndexRegistry.sol:130]: done

File: src/contracts/middleware/IndexRegistry.sol (hunk 2 above)
Comment (Sidu28) [reply src/contracts/middleware/IndexRegistry.sol:41]: done

File: src/contracts/interfaces/IIndexRegistry.sol (hunk 1 above)
Comment (Sidu28) [reply src/contracts/interfaces/IIndexRegistry.sol:38]: done

File: src/contracts/middleware/IndexRegistry.sol (hunk 6 above)
Comment (Sidu28) [reply src/contracts/middleware/IndexRegistry.sol:65]: added

File: src/contracts/middleware/IndexRegistry.sol (hunk 5 above)
Comment (Sidu28) [reply src/contracts/middleware/IndexRegistry.sol:190]: https://github.com/Layr-Labs/eigenlayer-contracts/blob/42234da18a78b710b1fa9a695820d4a6c06edf9e/src/contracts/middleware/IndexRegistry.sol#L83

File: src/contracts/middleware/IndexRegistry.sol [hunk 7]
```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity =0.8.12;
+
+
+import "../interfaces/IIndexRegistry.sol";
+import "../interfaces/IRegistryCoordinator.sol";
+import "../libraries/BN254.sol";
+import "forge-std/Test.sol";
```
Comment (gpsanant) [inline src/contracts/middleware/IndexRegistry.sol:8]: Test import

File: src/contracts/middleware/IndexRegistry.sol [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+     */
+    function registerOperator(bytes32 operatorId, bytes calldata quorumNumbers) external onlyRegistryCoordinator {
+        //add operator to operatorList
+        globalOperatorList.push(operatorId);
+
+        for (uint i = 0; i < quorumNumbers.length; i++) {
+            uint8 quorumNumber = uint8(quorumNumbers[i]);
+
+            //this is the would-be index of the operator being registered, the total number of operators for that quorum (which is last index + 1)
+            uint32 numOperators = totalOperatorsHistory[quorumNumber].length > 0 ? totalOperatorsHistory[quorumNumber][totalOperatorsHistory[quorumNumber].length - 1].index : 0;
```
Comment (gpsanant) [inline src/contracts/middleware/IndexRegistry.sol:53]: amortize length here

Comment (Sidu28) [reply src/contracts/middleware/IndexRegistry.sol]: fixed.  Also is amortize the right word here?

**Pull request merged**